### PR TITLE
CMR-4630: Modified the indication of all collection update.

### PR DESCRIPTION
**Pull request by sxu123**
In bulk update body, pass in :concept-ids ["ALL"] to indicate all the collections for the provider will be updated, rather than not passing any concept-ids in the previous implementation.

### CODE REVIEW

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj [hunk 1]
```diff
@@ -75,6 +75,18 @@
...
+            {:concept-ids ["all" ]
+             :name "TEST NAME 1"
+             :update-field "SCIENCE_KEYWORDS"
+             :update-type "ADD_TO_EXISTING"
+             :update-value {:Category "EARTH SCIENCE"
+                            :Topic "LAND SURFACE"
+                            :Term "SURFACE RADIATIVE PROPERTIES"
+                            :VariableLevel1 "REFLECTANCE"}}
+            400
             ["Concept-ids not found - None provided in the request, and none are associated with the provider-id either."]
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj]: Is this test saying the provider has no collections so specifying "ALL" returns an error?

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj [hunk 2]
```diff
@@ -75,6 +75,18 @@
                             :Term "SURFACE RADIATIVE PROPERTIES"
                             :VariableLevel1 "REFLECTANCE"}}
             400
+            ["object has missing required properties ([\"concept-ids\"])"]
+
+            "All concept-ids"
+            {:concept-ids ["all" ]
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj]: You test the blank spaces elsewhere. Should get rid of the blank space here.

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj [hunk 3]
```diff
@@ -188,14 +188,16 @@
         bulk-update-options1 {:token (e/login (s/context) "user1") :user-id "user2"}
         bulk-update-options2 {:token (e/login (s/context) "user1")}
         bulk-update-options3 {:user-id "user2"}
-        bulk-update-body {:update-type "ADD_TO_EXISTING"
+        bulk-update-body {:concept-ids ["ALL" ]
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj]: You test the blank spaces on line 199. Should get rid of the blank space here.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 4]
```diff
@@ -118,9 +118,9 @@
   "Get the concept-ids from either the concept-ids passed in or 
    from the provider. If both are empty, throws exception." 
   [context concept-ids provider-id]
-  (if-let [concept-ids (if (seq concept-ids)
-                         concept-ids
-                         (get-provider-concept-ids context provider-id))]
+  (if-let [concept-ids (if (= "[\"ALL\"]" (string/trim (string/upper-case concept-ids)))
```
Comment (oubiwann) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: I'm not the biggest fan of nested conditionals if there are obvious, idiomatic alternatives, but this is fairly small ... and straight-forward.

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj (hunk 1 above)
Comment (sxu123) [reply system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj]: Yes.

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj (hunk 2 above)
Comment (sxu123) [reply system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_endpoint_test.clj]: ok. done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 4 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Thanks. I will see if it looks better to have a separate function.

Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: This check looks off. It should just check `(map string/upper-case concept-ids)` equals `["ALL"]`. Also what if concept-ids is `["C1111-PROV1" "ALL"]`? I don't see this edge case being addressed.

Comment (chris-durbin) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: We should add a test for this case.

Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: I don't need to.  Honestly I thought about the case and when the user is mixing "ALL" with other concept-ids, it will be treated as the regular concept-ids. Only when the concept-ids = ["ALL"], it's treated as all collections.

Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: In which case, it will complain "ALL" concept-id doesn't exist.

Comment (ygliuvt) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: I think we should add validation of the concept-ids, any invalid concept-id will fail the request (including the mixed case). Only `["ALL"]` will be accepted as valid option for all collections.

Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 5]
```diff
@@ -114,13 +115,32 @@
...
+        concept-ids))))
+
 (defn- get-concept-ids
   "Get the concept-ids from either the concept-ids passed in or 
    from the provider. If both are empty, throws exception." 
   [context concept-ids provider-id]
-  (if-let [concept-ids (if (seq concept-ids)
-                         concept-ids
-                         (get-provider-concept-ids context provider-id))]
+  (if-let [concept-ids (get-validated-concept-ids context concept-ids provider-id)]
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: I think we should separate the validation from getting concept ids function. This way, we would have a definitive error message as to if there is no collection for the given provider or the provided concept ids list is empty.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 6]
```diff
@@ -114,13 +115,32 @@
...
+  [context concept-ids provider-id]
+  (if (= ["ALL"] (map string/upper-case concept-ids)) 
+    (get-provider-concept-ids context provider-id)
+    (let [err-msgs 
+           (for [id concept-ids
+                 :let [err-msg (concepts/concept-id-validation id)
+                       err-msg (if err-msg
+                                 err-msg
+                                 (when-not (string/starts-with? id "C")
+                                   [(str "Collection concept-id " id " does not start with C")]))]
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Perhaps the error message should read something like "Concept id [S1111-PROV1] is not a valid collection concept id, must start with C."

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 7]
```diff
@@ -114,13 +115,32 @@
...
+                 :when err-msg]
+             err-msg)]
+      (if (seq err-msgs)
+        (errors/throw-service-errors
+          :bad-request
+          [(string/join ", " err-msgs)])
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: use a `;` to separate the error messages.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 5 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 6 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 8]
```diff
@@ -106,26 +107,46 @@
...
+   concept-ids."
+  [concept-ids]
+  (when-not (= ["ALL"] (map string/upper-case concept-ids))
+    (let [err-msgs 
+           (for [id concept-ids
+                 :let [err-msg (concepts/concept-id-validation id)
+                       err-msg (if err-msg
+                                 err-msg
+                                 (when-not (string/starts-with? id "C")
+                                   [(str "Concept-id [" id
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: use `format` here would be better than `str`.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 9]
```diff
@@ -106,26 +107,46 @@
                                             update-type)]))))
 
 (defn- get-provider-concept-ids
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Rename this function to something like `get-provider-collection-concept-ids`.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 10]
```diff
@@ -106,26 +107,46 @@
...
-  "Returns a list of collection concept ids for a given provider-id"
+  "Returns a list of collection concept ids for a given provider-id.
+   Throws exception if non are returned."
   [context provider-id]
   (let [concepts (mdb/find-concepts context
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Since we can only update the latest revision of a collection, we should use `latest = true` to  only get the latest revision back. Also, I think this call will return tombstones (not certain) which we will want to skip.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 11]
```diff
@@ -106,26 +107,46 @@
...
+                                 (when-not (string/starts-with? id "C")
+                                   [(str "Concept-id [" id 
+                                         "] is not a valid collection concept id, "
+                                         "must start with C")]))]
+                 :when err-msg]
+             err-msg)]
+      (when (seq err-msgs)
+        (errors/throw-service-errors
+          :bad-request
+          [(string/join ", " err-msgs)])))))
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: We should use `;` as separate as there is `,` in some of the error messages.

Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 9 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 8 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Used str because the line is getting too long...   Will use a separate function then *Sign*

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 10 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done. I checked the metadata db call, it didn't exclude any deleted ones.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 12]
```diff
@@ -105,27 +106,52 @@
                                    [(format "A find value must be supplied when the update is of type %s"
                                             update-type)]))))
 
-(defn- get-provider-concept-ids
-  "Returns a list of collection concept ids for a given provider-id"
+(defn- get-provider-collection-concept-ids
+  "Returns a list of collection concept ids for a given provider-id.
+   Throws exception if non are returned."
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Change this line to "Raise error if no collections are found."

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 13]
```diff
@@ -105,27 +106,52 @@
...
   [context provider-id]
   (let [concepts (mdb/find-concepts context
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Change this to call `mdb/find-collections`.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 14]
```diff
@@ -105,27 +106,52 @@
...
+  "Returns a list of collection concept ids for a given provider-id.
+   Throws exception if non are returned."
   [context provider-id]
   (let [concepts (mdb/find-concepts context 
-                                    {:provider-id provider-id} 
-                                    :collection)]
-    (when (seq concepts)
-      (distinct (map :concept-id concepts)))))
+                                    {:provider-id provider-id
+                                     :latest "true"}
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: nit: `:latest true`.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 15]
```diff
@@ -105,27 +106,52 @@
...
-                                    {:provider-id provider-id} 
-                                    :collection)]
-    (when (seq concepts)
-      (distinct (map :concept-id concepts)))))
+                                    {:provider-id provider-id
+                                     :latest "true"} 
+                                    :collection)
+        concepts (remove :deleted concepts)]
+    (if (seq concepts)
+      (distinct (map :concept-id concepts))
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: `distinct` can be removed now.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 16]
```diff
@@ -105,27 +106,52 @@
...
+      (errors/throw-service-errors
+        :bad-request
+        ["No concept-ids are associated with the provider-id."]))))
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Change the error message to something like "There is no collections for provider xyz."

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 17]
```diff
@@ -105,27 +106,52 @@
...
+              err-msg 
+               (if err-msg
+                 err-msg
+                 (when-not (string/starts-with? id "C")
+                   [(format "Collection concept-id [%s] is invalid, must start with C" id)]))] 
+        :when err-msg]
+    err-msg))
+
+(defn- validate-concept-ids
+  "Throws exception if there exists invalid
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: nit: put the docstring in one line.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj [hunk 18]
```diff
@@ -105,27 +106,52 @@
...
+   concept-ids."
+  [concept-ids]
+  (when-not (= ["ALL"] (map string/upper-case concept-ids))
+    (let [err-msgs (get-collection-concept-id-validation-err-msgs concept-ids)]
+      (when (seq err-msgs)
+        (errors/throw-service-errors
+          :bad-request
+          [(string/join "; " err-msgs)])))))
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Sorry, we should change the `;` back to `,` as the individual errors are wrapped within brackets. Didn't see this earlier.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 13 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 14 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Both true and "true" work.  Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 15 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Done.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 16 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Changed to "There are no un-deleted collections for provider xyz."

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj [hunk 19]
```diff
@@ -860,6 +864,36 @@
       ;; revision-id not changed, format not changed, it's identical to the original-concepts.
       (is (= new-concepts original-concepts)))) 
 
+(deftest bulk-update-update-all-tomb-stone-test
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj]: nit: `tombstone` instead of `tomb-stone`.

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj [hunk 20]
```diff
@@ -860,6 +864,36 @@
...
+                          :update-value {:Category "EARTH SCIENCE"
+                                         :Topic "ATMOSPHERE"
+                                         :Term "AIR QUALITY"
+                                         :VariableLevel1 "EMISSIONS"}}]
+    ;; perform bulk update, verify that one collection is skipped because find-value is not found.
+    (let [response (ingest/bulk-update-collections "PROV1" bulk-update-body)
+          _ (index/wait-until-indexed)
+          collection-response (ingest/bulk-update-task-status "PROV1" (:task-id response))]
+      (is (= "COMPLETE" (:task-status collection-response)))
+      (is (= "Task completed with 1 SKIPPED out of 1 total collection update(s)." (:status-message collection-response))))
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj]: nit: it would be better to have this update succeed without any special messages. `skipped` is not as nice as total success.

Comment (sxu123) [reply system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj]: I only need to distinguish the case when there is a collection to be processed or not.

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 17 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Changed to "Raise error if there exists invalid concept-ids."

File: ingest-app/src/cmr/ingest/services/bulk_update_service.clj (hunk 18 above)
Comment (sxu123) [reply ingest-app/src/cmr/ingest/services/bulk_update_service.clj]: Okay...

File: system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj (hunk 19 above)
Comment (sxu123) [reply system-int-test/test/cmr/system_int_test/ingest/bulk_update/bulk_update_test.clj]: I'm confused. bulkupdate needs be separated, but tomb-stone need to be combined? Done.
